src/video/uikit/SDL_uikitpen.m

```diff
@@ -28,66 +28,160 @@
+        // it from -90 (back facing west) to 90 (back facing east).
+        const float xtilt = (180.0f - SDL_fabsf(azimuth_angle)) - 90.0f;
+
+        // the altitude_angle goes from 0 to 90 regardless of which direction the pen is lifting off the device, but SDL wants
+        // it from -90 (flat facing north) to 90 (flat facing south).
+        const float ytilt = (azimuth_angle < 0.0f) ? -(90.0f - altitude_angle) : (90.0f - altitude_angle);
+
+        // rotation is in radians, and only available on a later iOS.
+        const float rotation = rollAngle * radians_to_degrees;  // !!! FIXME: this might need adjustment, I don't have a pencil that supports it.
+
+        SDL_SendPenMotion(timestamp, penId, window, point->x, point->y);
+        SDL_SendPenAxis(timestamp, penId, window, SDL_PEN_AXIS_PRESSURE, pressure);
+        SDL_SendPenAxis(timestamp, penId, window, SDL_PEN_AXIS_XTILT, xtilt);
+        SDL_SendPenAxis(timestamp, penId, window, SDL_PEN_AXIS_YTILT, ytilt);
+        SDL_SendPenAxis(timestamp, penId, window, SDL_PEN_AXIS_ROTATION, rotation);
+        SDL_SendPenAxis(timestamp, penId, window, SDL_PEN_AXIS_DISTANCE, zOffset);
+    }
 }
 
-void UIKit_HandlePenMotion(SDL_uikitview *view, UITouch *pencil)
+#if !defined(SDL_PLATFORM_TVOS) && defined(__IPHONE_13_0)
+extern void UIKit_HandlePenHover(SDL_uikitview *view, UIHoverGestureRecognizer *recognizer)
 {
-    CGPoint point = [pencil locationInView:view];
-    SDL_SendPenMotion(UIKit_GetEventTimestamp([pencil timestamp]), penId, [view getSDLWindow], point.x, point.y);
-    SDL_SendPenAxis(UIKit_GetEventTimestamp([pencil timestamp]), penId, [view getSDLWindow], SDL_PEN_AXIS_PRESSURE, [pencil force] / [pencil maximumPossibleForce]);
-    NSLog(@"ALTITUDE: %f", [pencil altitudeAngle]);
-    NSLog(@"AZIMUTH VECTOR: %@", [NSValue valueWithCGVector: [pencil azimuthUnitVectorInView:view]]);
-    NSLog(@"AZIMUTH ANGLE: %f", [pencil azimuthAngleInView:view]);
-    // hold it
-    //    SDL_SendPenAxis(0, penId, [view getSDLWindow], SDL_PEN_AXIS_XTILT, [pencil altitudeAngle] / M_PI);
+    float zOffset = 0.0f;
+    if (@available(iOS 16.1, *)) {
+        zOffset = (float) [recognizer zOffset];
+    }
+
+    float azimuthAngleInView = 0.0f;
+    if (@available(iOS 16.4, *)) {
+        azimuthAngleInView = (float) [recognizer azimuthAngleInView:view];
+    }
+
+    float altitudeAngle = 0.0f;
+    if (@available(iOS 16.4, *)) {
+        altitudeAngle = (float) [recognizer altitudeAngle];
+    }
+
+    float rollAngle = 0.0f;
+    if (@available(iOS 17.5, *)) {
+        rollAngle = (float) [recognizer rollAngle];
+    }
+
+    SDL_Window *window = [view getSDLWindow];
+    const CGPoint point = [recognizer locationInView:view];
+
+    // force is zero here; if you're here, you're not touching.
+    // !!! FIXME: no timestamp on these...?
+    UIKit_HandlePenAxes(window, 0, zOffset, &point, 0.0f, 1.0f, azimuthAngleInView, altitudeAngle, rollAngle);
```

Comment:
You may be able to read the timestamp by fetching the `UITouch` representing the pencil, by `[recognizer locationOfTouch:0 inView:view]` then `[touch timestamp]`. 

I haven't tested if it works, but given that pen input is generally interpreted as `UITouch` with `UITouchTypePencil`, it might work well.

Reply:
It returns a CGPoint, not a UITouch, though...?

Reply:
Ah, my bad.